src/tqec/compile/compile.py

```diff
@@ -0,0 +1,282 @@
+from collections.abc import Iterable
+from dataclasses import dataclass
+from typing import Literal
+import cirq
+import stim
+import stimcirq
+
+from tqec.circuit.operations.measurement import Measurement
+from tqec.circuit.operations.operation import make_observable
+from tqec.circuit.operations.transformer import transform_to_stimcirq_compatible
+from tqec.circuit.schedule import ScheduledCircuit
+from tqec.compile.observables import (
+    get_center_qubit_at_horizontal_pipe,
+    get_midline_qubits_for_cube,
+    get_stabilizer_region_qubits_for_pipe,
+)
+from tqec.compile.substitute import (
+    SubstitutionKey,
+    SubstitutionRule,
+    DEFAULT_SUBSTITUTION_RULES,
+)
+from tqec.exceptions import TQECException
+from tqec.noise_models.base import BaseNoiseModel
+from tqec.plaquette.plaquette import RepeatedPlaquettes
+from tqec.position import Direction3D, Position3D
+from tqec.sketchup.block_graph import AbstractObservable, BlockGraph, Cube
+from tqec.compile.block import CompiledBlock, TiledBlocks
+from tqec.compile.specs import DEFAULT_SPEC_RULES, CubeSpec, SpecRule
+from tqec.templates.scale import round_or_fail
+
+
+@dataclass
+class CompiledGraph:
+    """Represents a compiled block graph.
+
+    This class should be easy to scale and generate circuits directly.
+    """
+
+    block_graph: BlockGraph
+    tiles_by_time: list[TiledBlocks]
```

Comment:
Maybe add an explicit `__post_init__` method to encode the pre-conditions on this? For example, `tiles_by_time` is required to have at least one entry, but you have to read some methods to know that. A `__post_init__` method is, in my opinion, a nice way to document the pre-conditions and ensure that they are indeed checked.